ArcGISRuntimeSDKQt_CppSamples/Routing/NavigateRoute/NavigateRoute.cpp

```diff
@@ -0,0 +1,271 @@
+  m_navigationButtonEnabled = false;
+  emit navigationButtonChanged();
+
+  // get the directions for the route
+  m_directions = m_route.directionManeuvers(this);
+
+  // create a route tracker
+  m_routeTracker = new RouteTracker(m_routeResult, 0, this);
+  connectRouteTrackerSignals();
+
+  // enable "recenter" button when location display is moved from nagivation mode
+  connect(m_mapView->locationDisplay(), &LocationDisplay::autoPanModeChanged, this, [this](LocationDisplayAutoPanMode autoPanMode)
+  {
+    m_recenterButtonEnabled = autoPanMode != LocationDisplayAutoPanMode::Navigation;
+    emit recenterButtonChanged();
+  });
+
+  connect(m_mapView->locationDisplay(), &LocationDisplay::locationChanged, this, [this](const Location& location)
+  {
+    Point projectedLocation = GeometryEngine::project(location.position(), SpatialReference::wgs84());
+    QGeoPositionInfo tempLocation(QGeoCoordinate(projectedLocation.y(), projectedLocation.x()), QDateTime::currentDateTime());
+    m_routeTracker->trackLocation(tempLocation);
+  });
+
+  // turn on map view's navigation mode
+  m_mapView->locationDisplay()->setAutoPanMode(LocationDisplayAutoPanMode::Navigation);
+
+  // add a data source for the location display
+  m_simulatedLocationDataSource = new SimulatedLocationDataSource(this);
+  m_simulatedLocationDataSource->setLocationsWithPolyline(m_route.routeGeometry());
+  m_mapView->locationDisplay()->setDataSource(m_simulatedLocationDataSource);
+  m_simulatedLocationDataSource->start();
+}
+
+void NavigateRoute::connectRouteTrackerSignals()
+{
+  connect(m_routeTracker, &RouteTracker::newVoiceGuidance, this, [this](VoiceGuidance* voiceGuidance)
+  {
+    m_speaker->say(voiceGuidance->text());
+  });
+
+  connect(m_routeTracker, &RouteTracker::trackingStatusChanged, this, [this](TrackingStatus* rawTrackingStatus)
+  {
+    auto trackingStatus = std::unique_ptr<TrackingStatus>(rawTrackingStatus);
+    QString textString("Route status: \n");
+    if (trackingStatus->destinationStatus() == DestinationStatus::NotReached || trackingStatus->destinationStatus() == DestinationStatus::Approaching)
+    {
+      textString += "Distance remaining: " + trackingStatus->routeProgress()->remainingDistance()->displayText() + " " +
+          trackingStatus->routeProgress()->remainingDistance()->displayTextUnits().pluralDisplayName() + "\n";
+      QTime time = QTime::fromMSecsSinceStartOfDay(trackingStatus->routeProgress()->remainingTime() * 60 * 1000); // convert time to milliseconds
+      textString += "Time remaining: " + time.toString("hh:mm:ss") + "\n";
+
+      // display next direction
+      if (trackingStatus->currentManeuverIndex() + 1 < dynamic_cast<DirectionManeuverListModel*>(m_directions)->directionManeuvers().length())
+      {
+        textString += "Next direction: "  + dynamic_cast<DirectionManeuverListModel*>(m_directions)->directionManeuvers()[trackingStatus->currentManeuverIndex() + 1].directionText() + "\n";
+      }
```

Comment:
I'd want to check if this dynamic_cast was successful before accessing it. You could also consider saving some performance by just doing the cast once here and storing a variable. Maybe even just store `m_directions` itself as a `DirectionManeuverListModel` and cast it once when you acquire it.